gcp_variant_transforms/transforms/extract_variant_header.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2017 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A PTransform to output a PCollection of single ``HeaderFields`` extracted
+from ``Variant`` recoreds. This ``HeaderFields`` contains only fields (FORMAT,
+INFO) not already defined in the header section of the VCF files."""
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+
+from vcf.parser import _Format as Format
+from vcf.parser import _Info as Info
+from vcf.parser import field_counts
+
+from gcp_variant_transforms.beam_io import vcf_header_io
+from gcp_variant_transforms.transforms import merge_headers
+
+
+__all__ = ['ExtractHeaderFieldsFromVariant']
+
+
+_FLAG = 'Flag'
+_STRING = 'String'
+_NO_DESCRIPTION = 'no_description'
+_UNKNOWN_SOURCE = 'unknown_source'
+_UNKNOWN_VERSION = 'unknown_version'
+
+
+class _ExtractHeaderFieldsFromVariant(beam.DoFn):
+  """Extracts header fields from ``Variant`` record."""
+
+  def _is_defined_info_field(self, field_key, headers):
+    return headers and field_key in headers.infos
+
+
+  def _is_defined_format_field(self, field_key, headers):
+    return headers and field_key in headers.formats
+
+
+  def _get_field_counts(self, field_value):
+    # TODO(nmousavi): return filed count by conditioning on field_value type
+    # (list, bool, o/w) once variant_to_bigquery code can deduct type for
+    # missing header fields from extracted final headers. Currently it inserts
+    # fields with missing header to bigquery as a repeated field.
+    return field_counts['.']
+
+
+  def _get_field_type(self, field_value):
+    if isinstance(field_value, bool):
+      return _FLAG
+    else:
+      return _STRING
+
+
+  def _extract_info_fields(self, variant, headers):
+    infos = {}
+    for info_field_key, variant_info in variant.info.iteritems():
+      info_field_value = variant_info.data
+      if not self._is_defined_info_field(info_field_key, headers):
```

Comment:
nit: consider changing to `contains_info_field(info_field_key, defined_header.infos)` or even just `contains_field` if defined_header cannot be none.

Reply:
Changed it, not relevant anymore.